drivers/sensor/mhz19b/mhz19b.c

```diff
@@ -0,0 +1,323 @@
+
+#include "mhz19b-priv.h"
+
+static void mhz19b_uart_flush(const struct device *uart_dev)
+{
+	uint8_t c;
+
+	while (uart_fifo_read(uart_dev, &c, 1) > 0) {
+		continue;
+	}
+}
+
+static uint8_t mhz19b_checksum(const uint8_t *data)
+{
+	uint8_t cs = 0;
+
+	for (uint8_t i = 1; i < MHZ19B_BUF_LEN - 1; i++) {
+		cs += data[i];
+	}
+
+	return 0xff - cs + 1;
+}
+
+static int mhz19b_send_cmd(const struct device *dev, const enum mhz19b_cmd_idx cmd_idx,
+			   const bool has_rsp)
+{
+	struct mhz19b_data *data = dev->data;
+	const struct mhz19b_cfg *cfg = dev->config;
+	int ret;
+
+	/* Make sure last command has been transferred */
+	ret = k_sem_take(&data->tx_sem, MHZ19B_WAIT);
+	if (ret) {
+		return ret;
+	}
+
+	data->cmd_idx = cmd_idx;
+	data->has_rsp = has_rsp;
+	k_sem_reset(&data->rx_sem);
+
+#ifdef CONFIG_MHZ19B_KEEP_MCU_AWAKE
+	if (has_rsp) {
+		pm_constraint_set(PM_STATE_SUSPEND_TO_IDLE);
+	}
+#endif
+
+	uart_irq_tx_enable(cfg->uart_dev);
+
+	if (has_rsp) {
+		uart_irq_rx_enable(cfg->uart_dev);
+		ret = k_sem_take(&data->rx_sem, MHZ19B_WAIT);
+#ifdef CONFIG_MHZ19B_KEEP_MCU_AWAKE
+		pm_constraint_release(PM_STATE_SUSPEND_TO_IDLE);
+#endif
+	}
+
+	return ret;
+}
+
+static inline int mhz19b_send_config(const struct device *dev, const enum mhz19b_cmd_idx cmd_idx)
+{
+	struct mhz19b_data *data = dev->data;
+	int ret;
+
+	ret = mhz19b_send_cmd(dev, cmd_idx, true);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (data->rd_data[MHZ19B_RX_CMD_IDX] != mhz19b_cmds[data->cmd_idx][MHZ19B_TX_CMD_IDX]) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static inline int mhz19b_poll_data(const struct device *dev, const enum mhz19b_cmd_idx cmd_idx)
+{
+	struct mhz19b_data *data = dev->data;
+	uint8_t checksum;
+	int ret;
+
+	ret = mhz19b_send_cmd(dev, cmd_idx, true);
+	if (ret < 0) {
+		return ret;
+	}
+
+	checksum = mhz19b_checksum(data->rd_data);
+	if (checksum != data->rd_data[MHZ19B_CHECKSUM_IDX]) {
+		data->data_valid = false;
+		return -EBADMSG;
+	}
+
+	data->data_valid = true;
+
+	switch (cmd_idx) {
+	case MHZ19B_CMD_IDX_GET_CO2:
+		data->data = sys_get_be16(&data->rd_data[2]);
+		break;
+	case MHZ19B_CMD_IDX_GET_RANGE:
+		data->data = sys_get_be16(&data->rd_data[4]);
+		break;
+	case MHZ19B_CMD_IDX_GET_ABC:
+		data->data = (uint16_t)data->rd_data[7];
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int mhz19b_channel_get(const struct device *dev, enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	struct mhz19b_data *data = dev->data;
+
+	if (chan != SENSOR_CHAN_CO2) {
+		return -ENOTSUP;
+	}
+
+	if (!data->data_valid) {
+		LOG_DBG("Checksum mismatch");
+		return -EINVAL;
+	}
+
+	val->val1 = (int32_t)data->data;
+	val->val2 = 0;
+
+	return 0;
+}
+
+static int mhz19b_attr_set(const struct device *dev, enum sensor_channel chan,
+			   enum sensor_attribute attr, const struct sensor_value *val)
+{
+	if (chan != SENSOR_CHAN_CO2) {
+		return -ENOTSUP;
+	}
+
+	switch (attr) {
+	case SENSOR_ATTR_FULL_SCALE:
+		if (val->val1 == 2000) {
```

Comment:
Maybe let's try to get rid of those conditionals?

<pre>
static enum mhz19b_cmd_idx val_to_mhz19b_cmd_idx(s32_t val) 
{
    switch(val) {
        case 0:
            return MHZ19B_CMD_IDX_SET_ABC_OFF;
        case 1:
            return MHZ19B_CMD_IDX_SET_ABC_ON;
        case 2000:
            return MHZ19B_CMD_IDX_SET_RANGE_2000;
        case 5000:
            return MHZ19B_CMD_IDX_SET_RANGE_5000;
        case 10000:
            return MHZ19B_CMD_IDX_SET_RANGE_10000;
        default:
            return MHZ19B_CMD_IDX_INVALID;
    }
}

static int mhz19b_attr_set(const struct device *dev, enum sensor_channel chan,
			   enum sensor_attribute attr, const struct sensor_value *val)
{
    enum mhz19b_cmd_idx cmd_idx;

	if (chan != SENSOR_CHAN_CO2) {
		return -ENOTSUP;
	}

	switch (attr) {
	case SENSOR_ATTR_FULL_SCALE:
	    __fallthrough;
	case SENSOR_ATTR_MHZ19B_ABC:
	    cmd_idx = val_to_mhz19b_cmd_idx(val->val1);
	    if (cmd_idx == MHZ19B_CMD_IDX_INVALID) {
	        return -ENOTSUP;
	    }
	    return mhz19b_send_config(dev, cmd_idx);
	default:
		return -ENOTSUP;
	}
}
</pre>

Reply:
Thanks for the suggestion, but I think `SENSOR_ATTR_FULL_SCALE` should be separated from `SENSOR_ATTR_MHZ19B_ABC`. I think I can use switch instead of if-else though, for slight performance improvement.

Reply:
You are using the same `mhz19b_send_config` for both cases. The example above could be even more improved:

<pre>
switch (attr) {
	case SENSOR_ATTR_FULL_SCALE:
	    __fallthrough;
	case SENSOR_ATTR_MHZ19B_ABC:
	    cmd_idx = val_to_mhz19b_cmd_idx(val->val1);
	    if (cmd_idx != MHZ19B_CMD_IDX_INVALID) {
	        return mhz19b_send_config(dev, cmd_idx);
	    }
	    __fallthrough;
	default:
		return -ENOTSUP;
	}
</pre>

so you can get rid of the second `return -ENOTSUP;` line.

Reply:
Thanks for the suggestions, I've created internal functions to handle each of the `attr` so that it looks a bit cleaner. A small bonus to this is that now I can use the internal functions to configure the mhz19b on init directly without creating a temporary `struct sensor_value` var.

I didn't use your suggested code because I think it makes the driver less straightforward to read.